pilot/pkg/networking/core/v1alpha3/route/route.go

```diff
@@ -138,8 +148,12 @@ func separateVSHostsAndServices(virtualService config.Config,
 		for svcHost, svc := range serviceRegistry {
 			// *.foo.global matches *.global
 			if svcHost.Matches(hostname) {
+				// Change the host name of service to the host name in virtual service as we have to build domains based
+				// on virtual service host names.
+				svc.Hostname = hostname
```

Comment:
is mutation ok here? Is this not some struct that is shared beyond this function?

Reply:
looks not ok, it is a pointer and shared by all proxies

Reply:
It is safe to mutate. Because we build new service objects here https://github.com/istio/istio/blob/dbf6a9e1c123bf6abbab45683b1c6c1f7660916c/pilot/pkg/networking/core/v1alpha3/httproute.go#L220

Reply:
The other case listenerPort=0, it is not reconstructed

Reply:
I have copied there also

Reply:
What if we have `*.foo.global` and `*.global`? Both will match the `svc`, and then try to mutate the Hostname?

Reply:
This is no different from what we already do today. We pick first matching service. Here we are picking first matching host in VS. So should be fine.

Reply:
Handled this and added in https://github.com/istio/istio/pull/34137 as this PR looks like might need more discussion.